### ISSUE TITLE
gzip module gzips images

### ISSUE THREAD

**Issue by brk0v** [mholt/caddy#108]
gzip module gzips images.

Example:
```
$ curl -H "Accept-Encoding: gzip" https://caddyserver.com/resources/images/caddy-boxed.png -o caddy-boxed.png
$ file caddy-boxed.png
caddy-boxed.png: gzip compressed data

$ curl https://caddyserver.com/resources/images/caddy-boxed.png -o caddy-boxed.png
$ file caddy-boxed.png
caddy-boxed.png: PNG image data, 391 x 85, 8-bit colormap, non-interlaced
```

**Comment by mholt** [comment]
That's right. Is it not what you are expecting?

**Comment by ApfelUser** [comment]
@mholt Images should not be gzipped as they can get bigger and already are compressed.

From **Google Page Speed:**

> Don't use gzip for image or other binary files
>
> Image file formats supported by the web, as well as videos, PDFs and other binary formats, are already compressed; using gzip on them won't provide any additional benefit, and can actually make them larger. To compress images, optimize them.

**Comment by mholt** [comment]
Caddy's website compresses the images, but Page Speed doesn't report any errors there - says it passed the compression test, in fact. What's the link to the page where you got that info? I'd like to look into it some more.

Incidentally, Page Speed reminds me that we need to look into cache-related headers like Cache-Control and E-Tag.

**Comment by ApfelUser** [comment]
Yes, PageSpeed Insights doesn't complain. It seems like it is not looking for gzipped images. But images should still not be compressed as this [StackExchange](http://webmasters.stackexchange.com/questions/8382/gzipped-images-is-it-worth) thread shows. You could also implement something similar to the **NGINX** `gzip_types` directive, so that everyone can decide himself on what he wants to be compressed, which would be my favorite. Something which also would be cool would be if you could add support for static pre-compressed files so that people can use e.g. [zopfli](https://github.com/google/zopfli), which is what I am currently doing with NGINX.

**Comment by ApfelUser** [comment]
The Google PageSpeed reference doesn't exist anymore by the way. I think that is due to their [developer.google.com](https://developer.google.com) material design overhaul. A lot of good references just don't exist now anymore -.-

But you can see on the StackExchange thread that also e.g. Yahoo! has recommended **not** to compress images.

**Comment by mholt** [comment]
Thanks for the links and suggestions! You both are right. Gzip needs some more controls, and with them, a set of sensible defaults (like, for example, not compressing images) so the user can get optimal behavior without having to think about it.

How does your nginx config look for gzip_types and with zopfli? Just curious.

If we expand the gzip middleware, perhaps it'll be time to rename it from "gzip" and just call it "compress"...

**Comment by ApfelUser** [comment]
My gzip config looks like this:

```
gzip on;
gzip_disable "msie6";
gzip_static on;
gzip_vary on;
gzip_proxied any;
gzip_comp_level 6;
gzip_buffers 16 8k;
gzip_http_version 1.1;
gzip_types text/plain text/css application/json application/javascript application/x-javascript text/javascript text/xml application/xml application/rss+xml application/atom+xml application/rdf+xml application/xml+rss image/svg+xml;
```

If I e.g. look up `domain.tld/main.css`, **NGINX** is looking for `<root>/main.css.gz`. And the CSS is just pre-compressed with zopfli. If there is no `.gz`-file **NGINX** is just going to compress it on its own if it matches one of the **MIME-Types** in `gzip_types`. It is of course not using zopfli for that because that isn't built in. 

I thought, "maybe build zopfli right into caddy :smile: " and searched for a go version of zopfli. I actually also found one: [foobaz/go-zopfli](https://github.com/foobaz/go-zopfli), but the developer said that his "go port is 2-3 times slower than the C version, and compresses at approximately 100 kB/s.". And the C version is according to Wikipedia already pretty slow: "Using the default setting, it achieves typically about 5 % better compression than zlib but takes around 80 times longer"

So. Due to the slowness I would **not** recommend building zopfli into Caddy. At least not as the default compression engine. Maybe keep zlib or whatever you are using for compression and create a configuration option where people can decide between

```
compression zopfli
```

and

```
compression zlib
```

What do you think? Sorry for my late response but I live in UTC+1 :smile:

**Comment by abiosoft** [comment]
Unlike @mholt, I'm also on UTC+1 :).

What is pretty straight forward to implement now is `gzip_types` and  `gzip_disable` which I assume are content type and user agent respectively.

gzip Caddyfile syntax will now look like this
```
gzip {
    types text/plain text/css application/json application/javascript
    disable msie6
}
```

Is this okay for a start ?

**Comment by ApfelUser** [comment]
Yeah, that looks great! Thanks for the fast response :smiley:

**Comment by abiosoft** [comment]
Cool. Should be ready later today or tomorrow.

**Comment by abiosoft** [comment]
I think I'll prefer to use more descriptive words. `mimes` or `mime_types` to `types` and `disable_ua` to `disable`. That makes it clearer that the former refers to MIME and the latter refers to user agents.

I guess the default behaviour when MIME types are not set is to accept all. That makes it compatible with the current behaviour.

**Comment by ApfelUser** [comment]
The default behavior of **NGINX** is to just compress `text/html` when nothing is set in the `gzip_types` option. But they allow __*__ in `gzip_types`for compressing everything.

[Source](http://nginx.org/en/docs/http/ngx_http_gzip_module.html#gzip_types)

**Comment by abiosoft** [comment]
That makes sense. I will go ahead with this unless @mholt has different opinion.

Thanks.

**Comment by abiosoft** [comment]
I'm basically done with the code.

I noticed `gzip_disable` on **nginx** currently uses regexp matching and `msie6` is a preset for `MSIE [4-6]\.`. I tried searching for other presets but could not find any other one apart from `msie6`.

The options for Caddy are either I allow the `msie6` preset, enable regexp or use only string matching.
@mholt is not a fan of regexp right now so I will wait for his take on this.

**Comment by mholt** [comment]
Wow, lots can happen while you snooze. :smile: @abiosoft Thanks for taking the initiative here!

@ApfelUser Thanks for sharing your config. I have a few thoughts before we finalize going forward.

How do you feel about specifying file extensions instead of MIME types? I'm much more a fan of file extensions because they're short - meaning less opportunity for typos and makes for a cleaner config file) and it's obvious from the request whether or not a file will be compressed. This way the user doesn't have to think about MIME types or know what they are -- and I must advocate for this kind of user experience with Caddy.

I also recommend that we drop the browser-specific settings (especially those that rely on User-Agent strings). There is the Accept-Encoding header to define whether the browser supports gzip. I'm not interested in building in support for any older, dying technologies like IE6. Ultimately, this means we drop the tentative `disable` property.

However, you did mention compression level in your config - this is quite easy to change in Go, so how about we add a `level` property to the gzip directive, which defaults to the highest compression level?

In the case of a "naked" gzip directive (one that doesn't open a block), we can default to compressing everything except for .jpg, jpeg, .png, .mov, .mp4, .wmv, and .ogg probably? Your `gzip_types` line is very long - it'd be nice to keep things tidy. Maybe it would be shorter to specify which file types to *not* compress rather than all the file types *to* compress. What do you think about this?

**Comment by ApfelUser** [comment]
I actually agree on all your points :+1:

Don't know why NGINX chose to use MIME types. Maybe because they are more accurate? But I think there should be no problems.

**Comment by mschoebel** [comment]
Regarding the use of file extensions instead of mime types: a backend server (configured via `proxy` in the Caddyfile) might return a large `application/json` response for `https://example.com/api/v1/users` (REST API endpoint, no file extension). I think there should be a possibilty to `gzip` this response. For other API endpoints the backend might return images.

Maybe provide both options?

**Comment by ApfelUser** [comment]
Yeah, both options would be useful. Especially in the case @mschoebel mentioned.

**Comment by alixaxel** [comment]
@mschoebel I had exactly the same though before reading this issue.

Extensions are nice and all, but it just feels wrong specifying i.e. `.php` as one of the *gzipable* extensions.

Like @lenovouser I also think `gzip_static` like behavior would be a great idea.

If it helps, here's the MIME types I compress with nginx (I looked into every single one of these):

 - https://github.com/alixaxel/halBox/blob/master/overlay/nginx/etc/nginx/conf.d/http_gzip.conf

And you can find a list of corresponding extensions on:

 - https://github.com/alixaxel/halBox/blob/master/overlay/nginx/etc/nginx/mime.types

Basically, it translates to the Caddy equivalent of:

```
.appcache
.atom
.css
.csv
.cur
.dtd
.eot
.htc
.htm
.html
.ico
.js
.json
.jsonp
.less
.manifest
.md
.mml
.otf
.owl
.owx
.rdf
.rss
.shtml
.svg
.torrent
.ttc
.ttf
.txt
.vcf
.vtt
.webapp
.xhtml
.xml
```
